Maple2.File.IO/Nif/NifDocument.cs

```diff
@@ -0,0 +1,225 @@
+            --versionStart;
+        }
+
+        for (int i = 0; i < 4; ++i) {
+            int versionEnd = versionStart;
+
+            while (versionEnd < headerStringLength && fileData[versionEnd] != '.')
+                ++versionEnd;
+
+            string versionText = Encoding.UTF8.GetString(fileData, versionStart, versionEnd - versionStart);
+
+            header.Version[i] = ushort.Parse(versionText);
+
+            versionStart = versionEnd + 1;
+        }
+
+        return index;
+    }
+
+    public NifBlock? GetBlock(int index) {
+        if (index == -1) {
+            return null;
+        }
+
+        if (blocks[index] is not null) {
+            return blocks[index];
+        }
+
+        string blockType = header.BlockTypes[header.BlockTypeIndices[index]];
+
+        NifBlock block = blockType switch {
+            "NiPhysXProp" => new NiPhysXProp(index),
+            "NiPhysXPropDesc" => new NiPhysXPropDesc(index),
+            "NiPhysXActorDesc" => new NiPhysXActorDesc(index),
+            "NiPhysXShapeDesc" => new NiPhysXShapeDesc(index),
+            "NiPhysXMeshDesc" => new NiPhysXMeshDesc(index),
+            _ => new NifBlock(blockType, false, index)
+        };
+
+        blocks[index] = block;
+
+        return block;
+    }
+
+    public T? GetBlock<T>(int index) {
+        if (index == -1) {
+            return default;
+        }
+
+        NifBlock? block = GetBlock(index);
+
+        if (block is T nifBlock) {
+            return nifBlock;
+        }
+
+        throw new InvalidCastException($"Invalid block handle <{typeof(T).Name}> referenced block {index}");
+    }
+
+    public T? ReadBlockRef<T>() {
+        int index = Reader.ReadInt32();
+
+        if (index != -1) {
+            return GetBlock<T>(index);
+        }
+
+        return default;
+    }
+
+    public List<T> ReadBlockRefList<T>() {
+        List<T> blocks = new List<T>();
+
+        int count = Reader.ReadInt32();
+
+        blocks.EnsureCapacity(count);
+
+        for (int i = 0; i < count; ++i) {
+            T? block = ReadBlockRef<T>();
+
+            if (block is null) {
+                throw new InvalidCastException($"Null/invalid block handle in block list <{typeof(T).Name}> of size {count}");
+            }
+
+            blocks.Add(block);
+        }
+
+        return blocks;
+    }
+
+    public string ReadString() {
+        uint stringIndex = Reader.ReadUInt32();
+
+        return header.Strings[stringIndex];
+    }
+
+    public void Reading(NifBlock block) {
+        ReadingBlock = block;
+    }
+
+    public bool Parse() {
+        int index = ReadHeaderString();
+
+        if (header.Version[0] < 30) {
+            Console.WriteLine(RelPath);
+            Console.WriteLine($"NIF version number too low: {header.HeaderString}. Parser is built for v30+"); // is it possible to do warnings?
```

Comment:
can we throw exceptions instead of logging

Reply:
That'd cause issues with the stock MS2 data though. None of the NIFs that are used are < v30, but the client contains stock Gamebryo NIFs that are v20 despite them not being referenced.

Reply:
the caller can handle the exceptions and just ignore the error

Reply:
Would that block parsing the nifs that come after it though?

Reply:
How so? `Parse()` is called multiple times

Reply:
Ah, does it save execution information & resume with yield return? I could tell it was doing something funky there, but never looked into how that worked. So how would exception handling for nexon's gamebryo leftovers look from the ingest side?

Reply:
It doesn't need to save execution information, you would have returned after anyways. (If you catch the exception it's practically the same as a return)

You'd need some way to differentiate between the different exceptions if you wanted to do case-by-case handling (without string matching)